File: Builder/Builder.cs
// 1. Producto
public class Reporte
{
    private List<string> _partes = new List<string>();

    public void Agregar(string parte)
    {
        this._partes.Add(parte);
    }

    public void Mostrar()
    {
        Console.WriteLine($"Reporte construido con partes: {string.Join(", ", _partes)}");
    }
}

// 2. Interfaz Builder
public interface IReporteBuilder
{
    void Reset();
    void BuildEncabezado();
    void BuildMetricasClave();
    void BuildFlujoCaja();
    Reporte GetResultado();
}

// 3. Concrete Builder
public class ReporteFinancieroBuilder : IReporteBuilder
{
    private Reporte _reporte = new Reporte();

    public ReporteFinancieroBuilder() { this.Reset(); }

    public void Reset() { this._reporte = new Reporte(); }

    public void BuildEncabezado() { this._reporte.Agregar("Encabezado (Fecha, Título)"); }
    public void BuildMetricasClave() { this._reporte.Agregar("Métricas de Viabilidad (VPN, TIR)"); }
    public void BuildFlujoCaja() { this._reporte.Agregar("Flujo de Caja Proyectado"); }

    public Reporte GetResultado()
    {
        Reporte resultado = this._reporte;
        this.Reset();
        return resultado;
    }
}

// 4. Director
public class Director
{
    private IReporteBuilder _builder;

    public IReporteBuilder Builder
    {
        set { this._builder = value; }
    }

    // Construye un reporte básico
    public void ConstruirReporteBasico()
    {
        this._builder.BuildEncabezado();
        this._builder.BuildMetricasClave();
    }

    // Construye un reporte completo
    public void ConstruirReporteCompleto()
    {
        this._builder.BuildEncabezado();
        this._builder.BuildMetricasClave();
        this._builder.BuildFlujoCaja();
    }
}

// Uso:
/*
Director director = new Director();
ReporteFinancieroBuilder builder = new ReporteFinancieroBuilder();
director.Builder = builder;

Console.WriteLine("Reporte Básico:");
director.ConstruirReporteBasico();
builder.GetResultado().Mostrar();

Console.WriteLine("\nReporte Completo:");
director.ConstruirReporteCompleto();
builder.GetResultado().Mostrar();
*/

File: 03 - Builder/Builder.js
// 1. Producto
class Reporte {
    constructor() {
        this.partes = [];
    }
    agregar(parte) {
        this.partes.push(parte);
    }
    mostrar() {
        console.log(`Reporte construido con partes: ${this.partes.join(', ')}`);
    }
}

// 2. Builder
class ReporteFinancieroBuilder {
    constructor() {
        this.reset();
    }
    reset() {
        this.reporte = new Reporte();
    }
    buildEncabezado() {
        this.reporte.agregar("Encabezado (Fecha, Título)");
    }
    buildMetricasClave() {
        this.reporte.agregar("Métricas de Viabilidad (VPN, TIR)");
    }
    buildFlujoCaja() {
        this.reporte.agregar("Flujo de Caja Proyectado");
    }
    getResultado() {
        const resultado = this.reporte;
        this.reset();
        return resultado;
    }
}

// 3. Director
class Director {
    constructor(builder) {
        this.builder = builder;
    }

    setBuilder(builder) {
        this.builder = builder;
    }

    // Construye un reporte básico
    construirReporteBasico() {
        this.builder.buildEncabezado();
        this.builder.buildMetricasClave();
    }

    // Construye un reporte completo
    construirReporteCompleto() {
        this.builder.buildEncabezado();
        this.builder.buildMetricasClave();
        this.builder.buildFlujoCaja();
    }
}

// Uso:
/*
const builder = new ReporteFinancieroBuilder();
const director = new Director(builder);

console.log("Reporte Básico:");
director.construirReporteBasico();
builder.getResultado().mostrar();

console.log("\nReporte Completo:");
director.construirReporteCompleto();
builder.getResultado().mostrar();
*/
